### PR TITLE
Beejones/presentation exchange

### PR DESCRIPTION
**Pull request by beejones**
**Problem:**
The SDK needs to support the presentation exchange protocol

**Solution:**
Add the presentation exchange protocol to Requestor

**Validation:**
New unit tests

**Type of change:**
- [x] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

**Risk**:
- [x] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

**Work Item links**:
https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1059130

**Documentation Links**:
https://identity.foundation/presentation-exchange

### CODE REVIEW

File: lib/ApiOidcRequest/RequestorBuilder.ts [hunk 1]
```diff
@@ -2,8 +2,22 @@
...
+import { Crypto, CryptoBuilder, JoseBuilder, IPayloadProtectionSigning, Requestor, IRequestorPresentationExchange, IRequestorAttestation, IRequestor } from '../index';
+
+/**
+ * Defines the presentation protcol
+ */
+export enum PresentationProtocol {
+  /**
+   * The presentation exchange protocol
+   */
+  presentationExchange,
```
Comment (symorton) [inline lib/ApiOidcRequest/RequestorBuilder.ts:14]: so we now support both protocols? Or after publishing new version, we will only support new Presentation Exchange version?

File: lib/ApiValidation/ValidatorBuilder.ts [hunk 2]
```diff
@@ -72,7 +72,8 @@ export default class ValidatorBuilder {
         idToken: new IdTokenTokenValidator(validatorOptions, <IExpectedIdToken> {type: TokenType.idToken, configuration: this._trustedIssuerConfigurationsForIdTokens}),
         verifiableCredential: new VerifiableCredentialTokenValidator(validatorOptions, <IExpectedVerifiableCredential> {type: TokenType.verifiableCredential, contractIssuers: this._trustedIssuersForVerifiableCredentials}),
         verifiablePresentation: new VerifiablePresentationTokenValidator(validatorOptions, this.crypto, <IExpectedVerifiablePresentation> {type: TokenType.verifiablePresentation, didAudience: this.crypto.builder.did}),
-        siopPresentation: new SiopTokenValidator(validatorOptions, <IExpectedSiop> {type: TokenType.siopPresentation, audience: this._audienceUrl}),
+        siopPresentationAttestation: new SiopTokenValidator(validatorOptions, <IExpectedSiop> {type: TokenType.siopPresentationAttestation, audience: this._audienceUrl}),
```
Comment (symorton) [inline lib/ApiValidation/ValidatorBuilder.ts:75]: oh yes so we will support both

File: lib/InputValidation/ValidationHelpers.ts [hunk 3]
```diff
@@ -564,18 +541,61 @@ export class ValidationHelpers {
...
+      }
+    } else {
+      // Get descriptor map
+      const descriptorMap: any[] = jp.query(validationResponse.payloadObject, `$.presentation_submission.descriptor_map.*`);
+      validationResponse.tokensToValidate = {};
+
+      for(let inx = 0; inx < descriptorMap.length; inx++) {
+        const item = descriptorMap[inx];
+        if (item.path) {
+          const tokenFinder = jp.query(validationResponse.payloadObject, item.path);
```
Comment (symorton) [inline lib/InputValidation/ValidationHelpers.ts:570]: jsonpath lib?

File: lib/InputValidation/ValidationQueueItem.ts [hunk 4]
```diff
@@ -79,7 +79,7 @@ export default class ValidationQueueItem {
    * Return false if the token still needs to be validated or the validation failed
    */
   public get result(): boolean {
-    return !this.isValidated &&  this._validationResult.result;
+    return this.isValidated &&  this._validationResult.result;
```
Comment (symorton) [inline lib/InputValidation/ValidationQueueItem.ts:82]: was this a bug before? Why take away the "!"?

File: lib/ApiOidcRequest/RequestorBuilder.ts (hunk 1 above)
Comment (beejones) [reply lib/ApiOidcRequest/RequestorBuilder.ts:14]: Both. It is the choice of the RP.

File: lib/InputValidation/ValidationQueueItem.ts (hunk 4 above)
Comment (beejones) [reply lib/InputValidation/ValidationQueueItem.ts:82]: This was a bug

File: lib/InputValidation/ValidationHelpers.ts (hunk 3 above)
Comment (beejones) [reply lib/InputValidation/ValidationHelpers.ts:570]: yes jp is jsonpath

**Pull request merged**